### ISSUE TITLE
Cypher query fails on 4.0

### ISSUE THREAD

**Issue by webtic** [neo4j/neo4j#12389]
- Neo4j version: 4.0
- Operating system: macOS 10.13.6
- API/Driver: localhost:7474

**Steps to reproduce**

While this query ran fine on 3.5 :

```
MATCH (e:LearningActivity) 
        OPTIONAL MATCH (e)-[:ICONIC_IMAGE]-(img)
        WITH collect(img)[0] as img, e
        OPTIONAL MATCH (e)-[:IS_OWNER]-(m:Member)
        WITH DISTINCT e {
            typecode: 'la',
            url: '/learning-activity/'+e.slug,
            activity: properties(e),
            upload_uuid: img.upload_uuid,
            gewone_uuid: img.uuid,
            icon: apoc.text.join(['/ua',img.path],'/'),
            image: CASE EXISTS(img.upload_uuid) WHEN true THEN apoc.text.join(['objects',img.upload_uuid,img.thumb],'/') ELSE apoc.text.join(['objects',img.uuid,img.thumb],'/') END, 
            x: img,
            has_image: count(img),
            image_url: CASE EXISTS(img.upload_uuid) WHEN true THEN apoc.text.join(['objects',img.upload_uuid,img.thumb],'/') ELSE apoc.text.join(['objects',img.uuid,img.thumb],'/') END, 
            owner: { name: m.name, uuid: m.uuid, login: m.login }
        } AS r 
        RETURN collect(r)

```
It fails on 4.0 with this error:

```class org.neo4j.values.storable.NoValue cannot be cast to class org.neo4j.values.virtual.VirtualNodeValue (org.neo4j.values.storable.NoValue and org.neo4j.values.virtual.VirtualNodeValue are in unnamed module of loader 'app')```

Removing the first OPTIONAL solves the issue, indicating that there are probably nodes without any image but internally that is handled less gracefully than it was in 3.5

**Comment by Hunterness** [comment]
Hi, thanks for reporting. We will look into it.

/Team Cypher

**Comment by bachmanm** [comment]
+1

My query is it helps:

```
MATCH (p:Person)
WITH p ORDER BY p.lastComputed LIMIT 500
SET p.lastComputed=timestamp()
WITH p
MATCH (pos:Positive)-[:HAS_TEST]->(posTest:Test {result:'POSITIVE'})
OPTIONAL MATCH p0=((p)-[r0:FAMILY]-(pos))
OPTIONAL MATCH p1=((p)-[r1:AD_HOC_CONTACT]-(pos)) WHERE duration.between(r1.lastContact,posTest.timestamp).days < 7
OPTIONAL MATCH p2=((p)-[r2:REGULAR_CONTACT]-(pos)) WHERE duration.between(r2.lastContact,posTest.timestamp).days < 7
OPTIONAL MATCH p3=((p)-[:REGULAR_CONTACT|AD_HOC_CONTACT*2..2]-(pos)) WHERE ALL (x in relationships(p3) WHERE duration.between(x.lastContact,posTest.timestamp).days < 7)
WITH p, count(p0) as cnt0, count(p1) as cnt1, count(p2) as cnt2, count(p3) as cnt3
WITH p, cnt0*50+cnt1*10+cnt2*5+cnt3*2 as score
SET p.score=score
```

**Comment by sherfert** [comment]
This has been fixed in 4.0.2
